### PR TITLE
[MKL-DNN] BatchNormRelu Fusion

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
BatchNormRelu Fusion while training.

Main changes for this feature are:
1)	Implement `mx.nd.contrib.BatchNormWithReLU` operator and place it in `src/operator/contrib/` directory;
2)	Add a parameter to gluon interface `nn.BatchNorm(fuse_relu=True, **kwargs)` to call `mx.nd.contrib.BatchNormWithReLU` instead of `mx.nd.BatchNorm`

For users who wants to use this feature, he needs to:
1)	Symbolic: call `mx.sym.contrib.BatchNormWithReLU` directly to replace `mx.sym.BatchNorm` and `mx.sym.Activation`;
2)	Gluon: Add layer `nn.BatchNorm(fuse_relu=True)` instead of `[nn.BatchNorm(), nn.Activation()]`

@pengzhao-intel @TaoLv 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by xinyu-intel** [comment]
@zhreshold @hetong007 Can you help take a review?

**Comment by zhreshold** [comment]
Not sure what's the best way in gluon frontend. Since many models comes with BN + Activation manually, we need to modify the code anyway with BNRelu fusion, then how about a gluon.nn.BatchNormRelu layer instead?

**Comment by xinyu-intel** [comment]
@zhreshold Yes, i think both are ok. But if we create a new one `gluon.nn.BatchNormRelu`, we may need to duplicate too much code from `gluon.nn.BatchNorm` and may need make duplicated changes later.

**Comment by zhreshold** [comment]
You can make you current implementation a base one. For normal bn, subclass and use with_relu=False, for new bn, subclass and use with_relu=True

**Comment by xinyu-intel** [comment]
okay, i'll take a try.

**Comment by xinyu-intel** [comment]
@mxnet-bot run ci [centos-cpu, unix-gpu, windows-gpu]

**Comment by xinyu-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by xinyu-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by xinyu-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by xinyu-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by pengzhao-intel** [comment]
@xinyu-intel  is this ready to merge?

**Comment by xinyu-intel** [comment]
@xinyu-intel it should be ok.
